### ISSUE TITLE
convert pytorch d2l.ai demos to JAX

### ISSUE THREAD

**Issue by murphyk** [probml/pyprobml#686]
In [vol 1](https://probml.github.io/pml-book/book1.html) of my book, I use several notebooks taken from the book [Dive Into Deep Learning](http://d2l.ai/). These are written in pytorch and stored at   https://github.com/probml/probml-notebooks/tree/main/notebooks-d2l.
The task is to convert these notebooks into JAX/FLAX code.
It's okay to keep using torch dataloaders, but you should replace `torch.tensor` with `jax.np.array`, and `torch.nn` with `flax.linen.nn`, etc. Be sure to replace idioms like `np.random.seed(seed=1); torch.manual_seed(1)` with  proper JAX PRNGs.

You should make a JAX version of each notebook, and open a PR for each one at 
https://github.com/probml/probml-notebooks/tree/main/notebooks-d2l.
If the original notebook is called `foo_torch.ipynb`, please call yours `foo_jax.ipynb`.

See [this example](https://github.com/probml/probml-notebooks/pull/46) for how to format your code and PR.

Before you start translating a notebook, please check that it has not already been done! (I only want one translation per notebook :)

- [x] [attention_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/attention_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/40
- [x] [batchnorm_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/batchnorm_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/60
- [x] [bert_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/bert_torch.ipynb)
- [x] [cnn1d_sentiment_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/cnn1d_sentiment_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/62
- [x] [conv2d_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/conv2d_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/52
- [x] [densenet_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/densenet_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/46
- [x] [entailment_attention_mlp_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/entailment_attention_mlp_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/75
- [x] [finetune_cnn_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/finetune_cnn_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/68
- [x] [gru_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/gru_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/66
- [x] [image_augmentation_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/image_augmentation_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/48
- [x] [lenet_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/lenet_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/42
- [x] [lstm_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/lstm_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/59
- [x] [multi_gpu_training_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/multi_gpu_training_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/77
- [x] [multi_head_attention.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/multi_head_attention.ipynb) - https://github.com/probml/probml-notebooks/pull/47
- [x] [naive_bayes_mnist.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/naive_bayes_mnist.ipynb) - https://github.com/probml/probml-notebooks/pull/52
- [x] [nmt_attention_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/nmt_attention_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/73
- [x] [nmt_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/nmt_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/72
- [x] [positional_encoding.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/positional_encoding.ipynb) - https://github.com/probml/probml-notebooks/pull/50
- [x] [resnet_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/resnet_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/45
- [x] [rnn_sentiment_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/rnn_sentiment_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/64
- [x] [rnn_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/rnn_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/67
- [x] [skipgram_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/skipgram_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/76
- [x] [transformers_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/transformers_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/71
- [x] [word_analogies_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/word_analogies_torch.ipynb) - https://github.com/probml/probml-notebooks/pull/41

**Comment by bicycleman15** [comment]
Hello 👋 
I would like to start working on converting `conv2d_torch.ipynb` to `conv2d_jax.ipynb` as a starting point.

**Comment by patel-zeel** [comment]
I have opened the following PR:

- [ ] (Added `attention_jax.ipynb` for `attention_torch.ipynb`)[https://github.com/probml/probml-notebooks/pull/40]

**Comment by zahramor** [comment]
Hi, I am interested to start with converting finetune_cnn_torch.ipynb to finetune_cnn_jax.ipynb.
Just one question, is it okay to keep torchvision. datasets and torchvision.transforms?

**Comment by arpitvaghela** [comment]
I have converted the following notebooks to jax will create a PR soon,
`conv2d_jax.ipynb`
`lenet_jax.ipynb` (with Animator)
`multi_head_attention_jax.ipynb`
`naive_bayes_mnist_jax.ipynb`
`densenet_jax.ipynb` (in progress)

**Comment by murphyk** [comment]
@zahramor 
> Hi, I would like to work on converting `finetune_cnn_torch.ipynb` to `finetune_cnn_jax.ipynb`, if still free. Just one question, is it okay to keep using `torchvision.datasets` and `torchvision.transforms`?

Yes for  `torchvision.datasets` (although in most cases there is a TFDS equivalent).
For transforms, it would be better to figure out a JAX equivalent (which may be faster).
Maybe [pix](https://github.com/deepmind/dm_pix) or some other library listed at https://github.com/n2cholas/awesome-jax?

**Comment by zahramor** [comment]
@murphyk Yes sure, no problem, I'll figure it out. Thanks for your replay :+1:

**Comment by susnato** [comment]
Hi,  I am currently working to convert [image_augmentation_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/image_augmentation_torch.ipynb) to image_augmentation_jax.ipynb .

**Comment by arihantbansal** [comment]
I'm working on [rnn_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/rnn_torch.ipynb)

**Comment by BalajiAI** [comment]
I'm working on converting [lstm_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/lstm_torch.ipynb) to lstm_jax.ipynb

**Comment by murphyk** [comment]
@patel-zeel It seems your checklist of PRs is out of date?

**Comment by patel-zeel** [comment]
> @patel-zeel It seems your checklist of PRs is out of date?

@murphyk I am updating it manually. Are there any missing notebooks or PRs?

**Comment by benlau6** [comment]
Hi, I am working on [transformers_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/transformers_torch.ipynb)

**Comment by nalzok** [comment]
I am working on [batchnorm_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/batchnorm_torch.ipynb).

**Comment by UmarJ** [comment]
Working on [rnn_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/rnn_torch.ipynb) and  [gru_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/gru_torch.ipynb) as well.

**Comment by UmarJ** [comment]
Working on [nmt_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/nmt_torch.ipynb) and  [nmt_attention_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/nmt_attention_torch.ipynb).

**Comment by yuanx749** [comment]
I am working on [entailment_attention_mlp_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/entailment_attention_mlp_torch.ipynb)

**Comment by susnato** [comment]
I am working on [bert_torch.ipynb](https://github.com/probml/probml-notebooks/blob/main/notebooks-d2l/bert_torch.ipynb)
